Capsule updates for coreboot+UefiPkgPayload

SergiiDmytruk opened the pull request:
# Description

The proposed changes are part of [UEFI Capsule Update for coreboot with EDK II](https://docs.dasharo.com/projects/capsule-updates/) project ([funded by NLnet foundation](https://nlnet.nl/project/UEFICapsuleUpdate/)), whose goal is to make updates of open source firmware based on coreboot and EDK2 easier for end users while also increasing update's security by limiting full flash access to the firmware itself.

Because capsules require persistent EFI variables, this also adds SMMSTOREv2 to EDK.

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - There is a possibility that building with `CAPSULE_SUPPORT` might require some changes.  Maybe some new changes in DSC file need to not apply if `CAPSULE_MAIN_FW_GUID` is empty.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - SMMSTORE and flash access in general might be considered related to security, but it's really dealt with by coreboot.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - No tests.

## How This Was Tested

A combination of coreboot changes from [here](https://review.coreboot.org/q/topic:%22uefi-capsules%22) and [here](https://review.coreboot.org/c/coreboot/+/82555) (for QEMU Q35) were combined with changes from this branch while update capsules were built as specified in [this overview](https://docs.dasharo.com/kb/edk2-capsule-updates/#capsule-creation).  Then `CapsuleApp.efi` and produced capsule file were made available in the VM and used to perform a successful firmware update.

Essentially the same changes but in forks of coreboot (https://github.com/Dasharo/coreboot/pull/509) and EDK2 (https://github.com/Dasharo/edk2/pull/147) were verified both on QEMU Q35 and [MSI PRO Z690-A DDR4](https://docs.dasharo.com/unified/msi/overview/#pro-z690-a-wifi-ddr4) hardware.

### How to test for yourself

For convenience, I combined coreboot's capsule and Q35 changes in a single branch [here](https://github.com/Dasharo/coreboot/tree/up/uefi-capsules-q35) (also fixes an issue with QEMU addressed by [now abandoned Q35 patch](https://review.coreboot.org/c/coreboot/+/82698)) and this branch contains all needed EDK changes.  After building coreboot with the following defconfig:

```defconfig
CONFIG_OPTION_BACKEND_NONE=y
CONFIG_BOARD_EMULATION_QEMU_X86_Q35=y
CONFIG_EDK2_BOOT_TIMEOUT=5
CONFIG_CPU_QEMU_X86_TSEG_SMM=y
CONFIG_USE_UDK_202005_BINDING=y
CONFIG_VGA_TEXT_FRAMEBUFFER=y
CONFIG_RESOURCE_ALLOCATION_TOP_DOWN=y
CONFIG_DRIVERS_EFI_VARIABLE_STORE=y
CONFIG_DRIVERS_EFI_FW_INFO=y
CONFIG_PAYLOAD_EDK2=y
CONFIG_EDK2_REPO_CUSTOM=y
CONFIG_EDK2_REPOSITORY="https://github.com/Dasharo/edk2-upstream.git"
CONFIG_EDK2_TAG_OR_REV="origin/up/enable-capsule-updates"
CONFIG_EDK2_SERIAL_SUPPORT=y
```

```bash
cp defconfig .config
make olddefconfig
make -j$(nproc)
```

One should be able to run the image in QEMU:
```bash
qemu-system-x86_64 -M q35,smm=on -enable-kvm -m 512 \
    -drive if=pflash,format=raw,unit=0,file=build/coreboot.rom \
    -drive file=disk.qcow2,format=qcow2 \
    -serial stdio -nographic
```

Here `disk.qcow2` is supposed to contain a capsule and `CapsuleApp.efi` (see https://docs.dasharo.com/kb/edk2-capsule-updates/ for how to produce those).  I deliberately didn't go into every little detail here but can do if anyone will face issues testing this.

## Integration Instructions

Those willing to make use of the changes from coreboot apart from pulling in relevant commits should have these options on:
```
CONFIG_USE_UDK_202005_BINDING=y
CONFIG_DRIVERS_EFI_VARIABLE_STORE=y
CONFIG_DRIVERS_EFI_FW_INFO=y
```

On EDK side this translates into `-D CAPSULE_SUPPORT=TRUE` and `-D CAPSULE_MAIN_FW_GUID=<firmware guid>`, plus building `CapsuleApp.efi` and capsules as mentioned above with proper key management for signing the capsules.

SergiiDmytruk commented:
Any feedback, anybody?  It doesn't have to be a detailed review, a high-level evaluation would do for a start, e.g. consider the following questions:
1. Are there any objections to the idea in general?  If so, what are they?
2. Is this PR too big for a review?  If so, what should get in first?
3. Is this the right place for the introduced functionality?  Should anything be moved to `edk2-platforms`?

shuoliu0 commented:
> Any feedback, anybody? It doesn't have to be a detailed review, a high-level evaluation would do for a start, e.g. consider the following questions:
> 
> 1. Are there any objections to the idea in general?  If so, what are they?
> 2. Is this PR too big for a review?  If so, what should get in first?
> 3. Is this the right place for the introduced functionality?  Should anything be moved to `edk2-platforms`?

Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.

shuoliu0 commented:
> > Any feedback, anybody? It doesn't have to be a detailed review, a high-level evaluation would do for a start, e.g. consider the following questions:
> > 
> > 1. Are there any objections to the idea in general?  If so, what are they?
> > 2. Is this PR too big for a review?  If so, what should get in first?
> > 3. Is this the right place for the introduced functionality?  Should anything be moved to `edk2-platforms`?
> 
> Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.

Another questions comes that, does this support to

shuoliu0 commented:
> > > Any feedback, anybody? It doesn't have to be a detailed review, a high-level evaluation would do for a start, e.g. consider the following questions:
> > > 
> > > 1. Are there any objections to the idea in general?  If so, what are they?
> > > 2. Is this PR too big for a review?  If so, what should get in first?
> > > 3. Is this the right place for the introduced functionality?  Should anything be moved to `edk2-platforms`?
> > 
> > 
> > Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.
> 
Another question is will the feature support to set capsule in an OS environment instead of UEFI shell? (maybe needs special tools).

P.S. sorry that I mistakenly closed the PR, just re-open it to continue the discussion.

SergiiDmytruk commented:
> Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.

Thanks for the suggestion, can do that.

> Another question is will the feature support to set capsule in an OS environment instead of UEFI shell? (maybe needs special tools).

Since it's the same EFI runtime API an operating system can submit an update capsule as long as its kernel can interact with capsule services of EFI.  In particular, it was tested with `cat firmware.cap > /dev/efi_capsule_loader` on Linux.

shuoliu0 commented:
> > Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.
> 
> Thanks for the suggestion, can do that.
> 
> > Another question is will the feature support to set capsule in an OS environment instead of UEFI shell? (maybe needs special tools).
> 
> Since it's the same EFI runtime API an operating system can submit an update capsule as long as its kernel can interact with capsule services of EFI. In particular, it was tested with `cat firmware.cap > /dev/efi_capsule_loader` on Linux.

Thanks for commenting. BTW, it looks like that if the coreboot.rom needs to be updated, still flashrom need to be used, right?

SergiiDmytruk commented:
> BTW, it looks like that if the coreboot.rom needs to be updated, still flashrom need to be used, right?

If you don't have an update in a capsule format, need to use `flashrom` or another non-UEFI method, because one can't pass raw flash image to `/dev/efi_capsule_loader` or EFI services.

SergiiDmytruk commented:
> > Maybe to split this PR into 2, one is add SMMSTORE support in UefiPayloadPkg, while another is for capsule update. The first one is generic to improve UefiPayloadPkg's interoperation with coreboot, which could be merged first.
> 
> Thanks for the suggestion, can do that.

Sent https://github.com/tianocore/edk2/pull/10902. Moved those commits to the start here, so this PR effectively includes the new one.

shuoliu0 commented on UefiPayloadPkg/Library/SblParseLib/SblParseLib.inf:
```diff
@@ -41,6 +41,8 @@
   gEfiGraphicsDeviceInfoHobGuid
   gUniversalPayloadPciRootBridgeInfoGuid
+  gUniversalPayloadSmbiosTableGuid
+  gUniversalPayloadAcpiTableGuid
```
Could you please add in commit message where the GUIDs are consumed?

P.S. the EDK2 building system will not check the GUIDs in inf, right? (the declaration is just for information?)

SergiiDmytruk replied:
In `SblParseLib.c` next to this file:
https://github.com/tianocore/edk2/blob/9039466cb68b8b41bee1aa6200faa3bfc8ffd145/UefiPayloadPkg/Library/SblParseLib/SblParseLib.c#L130
https://github.com/tianocore/edk2/blob/9039466cb68b8b41bee1aa6200faa3bfc8ffd145/UefiPayloadPkg/Library/SblParseLib/SblParseLib.c#L158

https://github.com/tianocore/edk2/pull/6026/commits/8180faa5d17d1bce212f8c4e9f161ce84d7e3fad results in `BlParseLib` being always built, which must be the reason behind missing GUIDs causing trouble for this PR.

> P.S. the EDK2 building system will not check the GUIDs in inf, right? (the declaration is just for information?)

It also affects declaration of symbols.  You can see undefined references [in CI output](https://dev.azure.com/tianocore/edk2-ci/_build/results?buildId=179009&view=logs&j=1108696c-b18a-53f7-19eb-f2a6129edcd7&t=dcd92f39-14c5-5715-2cce-332c64e33bac&l=39).  It might seem that declaring GUIDs is optional because they likely play a role only when linking executable modules.  Such modules typically include many libraries and as long as each used GUID is declared in at least one INF file the code will link successfully.  So it is easy to forget to add a GUID and not notice it because a declaration somewhere else brings it into the final image anyway. (I don't know all the details of tooling, but that's my impression of how it works.)

shuoliu0 replied:
Thanks!

shuoliu0 reviewed (commented):
It would be helpful to add 1-2 examples of the FW update cases, how fw info and CV info are  exposed and handled by the FMP protocol.

shuoliu0 commented on UefiPayloadPkg/Library/SmmStoreLib/SmmStore.c:
```diff
@@ -224,7 +276,62 @@ SmmStoreLibWrite (
+    return EFI_NO_MEDIA;
+  }
+
+  return WriteBlock (Lba, Offset, NumBytes, Buffer, SMMSTORE_CMD_USE_FULL_FLASH | SMMSTORE_CMD_RAW_WRITE);
```
BTW, for SMMSTORE_CMD_USE_FULL_FLASH, is there a plan to upstream at coreboot?

SergiiDmytruk replied:
> For SMMSTORE_CMD_USE_FULL_FLASH, the usable LBA is exceeding the smm store region?

Yes, in this case the base for addressing blocks is the start of the flash chip, not the start of the SMMSTORE region.

> BTW, for SMMSTORE_CMD_USE_FULL_FLASH, is there a plan to upstream at coreboot?

It has landed recently: https://review.coreboot.org/c/coreboot/+/83424

shuoliu0 replied:
Thanks, close this open.

shuoliu0 commented:
> It would be helpful to add 1-2 examples of the FW update cases, how fw info and CV info are exposed and handled by the FMP protocol.

Per reading https://docs.dasharo.com/projects/capsule-updates/#expected-flow-of-update-process, my understanding - corresponding the unique fw info, either 1 (if merged) or multiple capsule volumes (kept unmerged) could be generated, and the capsule could be the whole flash image (since the SMMSTORE_CMD_USE_FULL_FLASH is used to seek the offsets in full flash), but I'm not sure. More confirmation is welcomed.

SergiiDmytruk commented:
> > It would be helpful to add 1-2 examples of the FW update cases, how fw info and CV info are exposed and handled by the FMP protocol.
> 
> Per reading https://docs.dasharo.com/projects/capsule-updates/#expected-flow-of-update-process, my understanding - corresponding the unique fw info, either 1 (if merged) or multiple capsule volumes (kept unmerged) could be generated, and the capsule could be the whole flash image (since the SMMSTORE_CMD_USE_FULL_FLASH is used to seek the offsets in full flash), but I'm not sure. More confirmation is welcomed.

Right, it's worth noting that the proposed implementation updates the entire flash chip.  For this to work correctly Intel ME needs to be disabled and all flash chip protections need to be lifted (rather, not applied) if coreboot finds any capsules during initialization.  It is done to allow updating ME as well and change chip's layout in a new firmware (e.g., to give more space to BIOS region).  This can be made more flexible, say, by parsing coreboot's FMAP and only updating some regions but it needs embedding more knowledge about specific firmware image via some library.

The overall process looks like this (could be a bit off since some months have passed):
 1. `CONFIG_DRIVERS_EFI_MAIN_FW_GUID` is set to a GUID to be used in ESRT.  There are similar `CONFIG_DRIVERS_EFI_MAIN_FW_VERSION` and `CONFIG_DRIVERS_EFI_MAIN_FW_LSV` but they are less important.  See https://docs.dasharo.com/kb/capsule-updates-configs/ for more details.
 2. Capsules are created (see https://docs.dasharo.com/kb/edk2-capsule-updates/#capsule-creation).
 3. `BlParseLib` reads in those and `BlSupportDxe` publishes an ESRT entry to be parsed and exposed by operating systems or queried by EFI applications.
 4. A capsule is submitted to EDK via EFI runtime services.  It is left in RAM with special EFI variables `CapsuleUpdateData*` pointing to where they are.  A single variable can point to a single or multiple capsules because those can be chained.  The hard-limit in coreboot is randomly set to 32 individual capsules which is a really large number of updates.
 5. A warm reboot of the system is done (contents of RAM is largely preserved).
 6. coreboot checks for `CapsuleUpdateData`, collects capsules into a contiguous memory range and exposes its address and size via CBMEM.
 7. EDK checks (via `BlParseLib`) existence of coreboot capsules and creates HOBs for them.  It also sets boot mode to update mode.  Without coreboot this is done by PEI, most of the rest handling of capsules is done the same way regardless of whether coreboot is used or not because that code doesn't need to know what provides capsules.
 8. `PlatformBootManager` initiates dispatching of capsules upon detecting update mode.  `DxeCapsuleLibFmp` picks up capsules from HOBs and loads `FmpDxe.efi` included in capsules as a driver.  `FmpDxe` built against `FmpDeviceSmmLib`  registers `EFI_FIRMWARE_MANAGEMENT_PROTOCOL` on its initialization.  That protocol later gets invoked by `DxeCapsuleLibFmp` to perform flash update.

shuoliu0 commented:
> > > It would be helpful to add 1-2 examples of the FW update cases, how fw info and CV info are exposed and handled by the FMP protocol.
> > 
> > 
> > Per reading https://docs.dasharo.com/projects/capsule-updates/#expected-flow-of-update-process, my understanding - corresponding the unique fw info, either 1 (if merged) or multiple capsule volumes (kept unmerged) could be generated, and the capsule could be the whole flash image (since the SMMSTORE_CMD_USE_FULL_FLASH is used to seek the offsets in full flash), but I'm not sure. More confirmation is welcomed.
> 
> Right, it's worth noting that the proposed implementation updates the entire flash chip. For this to work correctly Intel ME needs to be disabled and all flash chip protections need to be lifted (rather, not applied) if coreboot finds any capsules during initialization. It is done to allow updating ME as well and change chip's layout in a new firmware (e.g., to give more space to BIOS region). This can be made more flexible, say, by parsing coreboot's FMAP and only updating some regions but it needs embedding more knowledge about specific firmware image via some library.
> 
> The overall process looks like this (could be a bit off since some months have passed):
> 
> 1. `CONFIG_DRIVERS_EFI_MAIN_FW_GUID` is set to a GUID to be used in ESRT.  There are similar `CONFIG_DRIVERS_EFI_MAIN_FW_VERSION` and `CONFIG_DRIVERS_EFI_MAIN_FW_LSV` but they are less important.  See https://docs.dasharo.com/kb/capsule-updates-configs/ for more details.
> 2. Capsules are created (see https://docs.dasharo.com/kb/edk2-capsule-updates/#capsule-creation).
> 3. `BlParseLib` reads in those and `BlSupportDxe` publishes an ESRT entry to be parsed and exposed by operating systems or queried by EFI applications.
> 4. A capsule is submitted to EDK via EFI runtime services.  It is left in RAM with special EFI variables `CapsuleUpdateData*` pointing to where they are.  A single variable can point to a single or multiple capsules because those can be chained.  The hard-limit in coreboot is randomly set to 32 individual capsules which is a really large number of updates.
> 5. A warm reboot of the system is done (contents of RAM is largely preserved).
> 6. coreboot checks for `CapsuleUpdateData`, collects capsules into a contiguous memory range and exposes its address and size via CBMEM.
> 7. EDK checks (via `BlParseLib`) existence of coreboot capsules and creates HOBs for them.  It also sets boot mode to update mode.  Without coreboot this is done by PEI, most of the rest handling of capsules is done the same way regardless of whether coreboot is used or not because that code doesn't need to know what provides capsules.
> 8. `PlatformBootManager` initiates dispatching of capsules upon detecting update mode.  `DxeCapsuleLibFmp` picks up capsules from HOBs and loads `FmpDxe.efi` included in capsules as a driver.  `FmpDxe` built against `FmpDeviceSmmLib`  registers `EFI_FIRMWARE_MANAGEMENT_PROTOCOL` on its initialization.  That protocol later gets invoked by `DxeCapsuleLibFmp` to perform flash update.

Thanks, it worthwhile to add comments in codes to mentioning some key points as above.

shuoliu0 commented on UefiPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -0,0 +1,1122 @@
+  // help, so they are employed only on successful flashing.
+  //
+
+  gRT->GetVariable         = GetVariableHook;
```
After setting new image, suppose the system will be reset? So the exposure window of the old RT services are actually short?

SergiiDmytruk replied:
The system will be reset shortly after this code is run regardless of the result.  The RT services will also be used right after this code is run to store result of applying each capsule (`Capsule####` variables).  I was looking into replacing FirmwareVolumeBlockProtocol used by variable services to let those variables be stored in the new image, but saw some static state in the relevant code which is not easy to change from the outside.

shuoliu0 replied:
Functionally, to link to stub and wait for another reset should be okay. But if below case exists, not sure if with the stub, some functionality issue will occur (maybe corner case but need to be addressed)?
```
The RT services will also be used right after this code is run to store result of applying each capsule (Capsule#### variables). 
```

SergiiDmytruk replied:
I don't think anything during processing of capsules or after it really depends on EFI variables, they are used to pass data across reboots.  Also the stubs are installed only on successful flashing after which the current firmware can be completely discarded.

shuoliu0 replied:
So actually these stubs are not invoked after a successful capsule update, right?

SergiiDmytruk replied:
They are invoked, but I don't think anything depends on results of invoking them.  [UEFI Specification](https://uefi.org/specs/UEFI/2.11/08_Services_Runtime_Services.html#uefi-variable-reporting-on-the-success-or-any-errors-encountered-in-processing-of-capsules-after-restart) doesn't suggest that setting the variables are important and code doesn't check result of setting them neither on error nor on success:

https://github.com/tianocore/edk2/blob/8aad683e59dc5342a4c9b8bc3d84133d4b1896f6/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c#L1350-L1357
https://github.com/tianocore/edk2/blob/8aad683e59dc5342a4c9b8bc3d84133d4b1896f6/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c#L1146-L1153
https://github.com/tianocore/edk2/blob/8aad683e59dc5342a4c9b8bc3d84133d4b1896f6/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c#L1566
https://github.com/tianocore/edk2/blob/8aad683e59dc5342a4c9b8bc3d84133d4b1896f6/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c#L1576
https://github.com/tianocore/edk2/blob/8aad683e59dc5342a4c9b8bc3d84133d4b1896f6/MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.c#L1589

shuoliu0 replied:
Maybe it would be helpful to mention in the commit message:

For a successful whole flash update, a reset will follow up and the variable access will not be accessible before the reset.

shuoliu0 replied:
The question would be for unsuccessful flash update:

1 - if a pre-update check fails, where the flash is not updated at all, the previous variable service could be used to record errors -> this is no problem.

2 - if a failure occurs in the middle of an update before fully success, the variable service is not hooked yet, will this be a problem?

shuoliu0 replied:
For 2, I guess it will be taken care of by IoError: label, right? Not sure if it could be better to directly halt the program there.

benjamindoron replied:
I wrote an FMAP parser for another feature we've been working on, and I can see if we're ready to upstream both the library and the feature in general, but I can say that an FMAP is a really easy structure to parse. Anyways, I think it's safe to skip it on the basis that changing the FMAP during a product's life cycle is possible, but unlikely.

SergiiDmytruk replied:
300 wasn't a random number, I basically imported coreboot's code for parsing FMAP in https://github.com/Dasharo/edk2/commit/5e022683e7b34d44f6ab76ea138ae95fa6cbfeb9. We use the library to move data from old to new firmware during an update, but it wasn't included in this PR as non-essential functionality.  I agree that layout changes are quite rare.

shuoliu0 replied:
Thanks all for the info. Personally I think an FMAP parser could be optional as long as it is commented clearly for the variable unavailable gap. While for the error happened in the middle of the flashing, there should be no way to mitigate so it should be logged clearly that user needs to do offline flash to rescue.

shuoliu0 replied:
Thanks

benjamindoron replied:
> 300 wasn't a random number, I basically imported coreboot's code for parsing FMAP in [Dasharo/edk2@5e02268](https://github.com/Dasharo/edk2/commit/5e022683e7b34d44f6ab76ea138ae95fa6cbfeb9). We use the library to move data from old to new firmware during an update, but it wasn't included in this PR as non-essential functionality. I agree that layout changes are quite rare.

coreboot passes the FMAP as yet another table, so you don't need all of that (the binary search). My code is up at https://github.com/benjamindoron/edk2/commit/f2f7f9170a29bf8e6b844eb8f19d85a71243e91c#diff-7de401b1c214f5f43e4bb1a8378c9dcd9c7cf5ecc6d217e39ece79b6daf1ada2

Anyways, I looked at FmpDxe, and once this returns all that it's doing is setting variables (which has no effect, per the hooks) and updating the progress for the user. Is the reason why you don't reset immediately because that would leave the user without a final indication of "success!" I can say that all callers would probably be fine with the platform rebooting at this time

SystemFirmwareReportDxe has a similar hooking implementation, so it seems like existing code agrees that this is good and workable

shuoliu0 commented on UefiPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -0,0 +1,1122 @@
+
+  //
+  // After the firmware on system flash was successfully updated working with
+  // variable store on the flash isn't safe.  Switch to the use of stubs which
```
For the security concern, could you please specify more?

SergiiDmytruk replied:
It's really about integrity rather than security.  SMMSTORE is overwritten and potentially moved within the flash chip, so using it until the reboot can accidentally corrupt the new image.  Expanded the comment: https://github.com/tianocore/edk2/compare/714555238024ffb08ae69c45b80ebb359b217706..2dc2f13efee5904c4339e80522c57748f5f0a245#diff-72cb1d4fadf58ff551d9bf949706e504d58c3482c99a778cee5224c21ac2ee16R1050

shuoliu0 replied:
I see, could you please update the commit message here?

SergiiDmytruk replied:
Updated commit message can be seen in https://github.com/tianocore/edk2/pull/6026/commits/d376d064bce2ca15558b23ad9ad169291f8e032c

SergiiDmytruk commented:
> it worthwhile to add comments in codes to mentioning some key points as above.

Improved messages of several commits to added more information and updated a couple comments where it's applicable (some other points are about coreboot and unmodified parts of EDK):
* https://github.com/tianocore/edk2/compare/714555238024ffb08ae69c45b80ebb359b217706..2dc2f13efee5904c4339e80522c57748f5f0a245#diff-72cb1d4fadf58ff551d9bf949706e504d58c3482c99a778cee5224c21ac2ee16
* https://github.com/tianocore/edk2/compare/714555238024ffb08ae69c45b80ebb359b217706..2dc2f13efee5904c4339e80522c57748f5f0a245#diff-2ebde4eb66fb05b7df6cbb40b5279056f0d219ab59499da1d9e48e2b118e0a2f

shuoliu0 commented on UefiPayloadPkg/UefiPayloadPkg.dsc:
```diff
@@ -952,6 +970,21 @@
+      # See BaseTools/Source/Python/Pkcs7Sign/Readme.md for more details on such
+      # PCDs and include files.
+      !include BaseTools/Source/Python/Pkcs7Sign/TestRoot.cer.gFmpDevicePkgTokenSpaceGuid.PcdFmpDevicePkcs7CertBufferXdr.inc
+  }
```
Per the comment block, it looks like there could be multiple FmpDxe, each target to update different capsule component. Here in this dsc, will all FmpDxe links to the same FmpDeviceLib here, a.k.a., FmpDeviceLib|UefiPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.inf?

If yes, maybe we need to make FmpDeviceLib per FmpDxe, since the added FmpDeviceLib only handle one case - the whole flash update.

SergiiDmytruk replied:
You're right, it makes more sense to declare it here: https://github.com/tianocore/edk2/compare/d376d064bce2ca15558b23ad9ad169291f8e032c..112352b2f40b319ed3a21c70855b528209008f22#diff-46ea4883eee85646f25a4e5732ffb1a8ee2447bffdf3aba462d0f9e1ab58e301

shuoliu0 replied:
Agree, thanks!

shuoliu0 commented on UefiPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -0,0 +1,1136 @@
+  // a recovery.  Doesn't seem that the calling code will do any of it.
+  //
+  FreePool (ReadBuffer);
+  DEBUG ((
```
It would be helpful to extend comment here for offline flash as metigation.

SergiiDmytruk replied:
https://github.com/tianocore/edk2/compare/112352b2f40b319ed3a21c70855b528209008f22..7165d8faa678bc3c32b88be307c9ae2b68e58645#diff-72cb1d4fadf58ff551d9bf949706e504d58c3482c99a778cee5224c21ac2ee16

shuoliu0 replied:
Thanks

benjamindoron commented on UefiPayloadPkg/UefiPayloadPkg.dsc:
```diff
@@ -316,13 +316,11 @@
   DebugAgentLib|MdeModulePkg/Library/DebugAgentLibNull/DebugAgentLibNull.inf
 !endif
   PlatformSupportLib|UefiPayloadPkg/Library/PlatformSupportLibNull/PlatformSupportLibNull.inf
-!if $(UNIVERSAL_PAYLOAD) == FALSE
```
The entrypoint module should turn bootloader-specific data formats into HOBs. However, that doesn't work for UPL yet, we need to extend coreboot and FituniversalPayloadEntry to produce and consume all the same information

This would only work on a UNIVERSAL_PAYLOAD build if coreboot produces coreboot tables, whereas once I get the https://review.coreboot.org/c/coreboot/+/84796 series merged (and either I or somebody else has coreboot pass all of its information along), UPL is supposed to use FDTs these days

SergiiDmytruk replied:
This has been undone: https://github.com/tianocore/edk2/compare/34f9d03df8220dbfbcda82aed163d988927ef7a3..a84de0dbf18d0802e2e06bc7057337ed9e77147b#diff-46ea4883eee85646f25a4e5732ffb1a8ee2447bffdf3aba462d0f9e1ab58e301R342

Now there is `FIRMWARE_INFO` HOB which provides information for ESRT and also version as a string used by FmpDeviceLib.  This slightly simplified changes to BlParser and FmpDeviceLib that no longer needs its own structure for maintaining the information.

benjamindoron reviewed (changes requested):
If SMMSTORE is going to be used as a generic write interface, then perhaps the SmmStoreLib declaration in the DSC should be brought out of the VARIABLE_SUPPORT condition. But that doesn't have to happen here/block merging for now.

Other than the use of BlParseLib in DXE, which I think we should address and reach a decision on if that should be allowed, this looks good to me.

SergiiDmytruk commented on UefiPayloadPkg/UefiPayloadPkg.dsc:
```diff
@@ -277,8 +277,10 @@
+  CapsuleLib|MdeModulePkg/Library/DxeCapsuleLibFmp/DxeCapsuleLib.inf
   BmpSupportLib|MdeModulePkg/Library/BaseBmpSupportLib/BaseBmpSupportLib.inf
+  # DisplayUpdateProgressLibGraphics aborts firmware update if GOP is missing, text progress works in more environments
+  DisplayUpdateProgressLib|MdeModulePkg/Library/DisplayUpdateProgressLibText/DisplayUpdateProgressLibText.inf
```
Added `!if $(BOOTSPLASH_IMAGE)`: https://github.com/tianocore/edk2/compare/34f9d03df8220dbfbcda82aed163d988927ef7a3..a84de0dbf18d0802e2e06bc7057337ed9e77147b#diff-46ea4883eee85646f25a4e5732ffb1a8ee2447bffdf3aba462d0f9e1ab58e301

SergiiDmytruk commented:
There were 2 force-pushes.  One to rebase PR and resolve conflicts and [another one](https://github.com/tianocore/edk2/compare/34f9d03df8220dbfbcda82aed163d988927ef7a3..a84de0dbf18d0802e2e06bc7057337ed9e77147b#diff-46ea4883eee85646f25a4e5732ffb1a8ee2447bffdf3aba462d0f9e1ab58e301) to perform requested changes without mixing changes from master.  Here is an overview of modifications:
 - `BlParser` is no longer used by `BlSupportDxe` or `FmpDeviceSmmLib`.  `UefiPayloadEntry` now publishes firmware details in `FIRMWARE_INFO` (`gEfiFirmwareInfoHobGuid`) HOB which gets consumed in DXE.
 - `BlParser` now forms version string as part of firmware information to not introduce 2 new functions for related pieces of information.
 - `FmpDeviceSmmLib` now uses `FIRMWARE_INFO` instead of a local equivalent.
 - Added `ASSERT (Hob != NULL);` to `UefiPayloadPkg/Library/PayloadEntryHobLib/Hob.c` for consistency with the other functions there.
 - Moved importing capsules to `BuildHobFromBl()` in `UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c` which is a function for building HOBs out of bootloader's data (haven't noticed it earlier).
 - Added `$(BOOTLOADER) == "COREBOOT"` to `UefiPayloadPkg/UefiPayloadPkg.dsc` to not enable things which won't work for Sbl.
 - Fixed building with `CAPSULE_SUPPORT=TRUE` but without `VARIABLE_SUPPORT=SMMSTORE` by declaring `SmmStoreLib` when either of the conditions is true.

benjamindoron commented:
I've tested this, and something about the `!include BaseTools/Source/Python/Pkcs7Sign/TestRoot.cer.gFmpDevicePkgTokenSpaceGuid.PcdFmpDevicePkcs7CertBufferXdr.inc` line for FmpDxe doesn't work, which means that the certificate that actually makes it inside the binary is the default of `{0}`. If I copy the contents of that file in-place of the `!include`, however, it does work as expected. Can you reproduce this?

On my system, it seems like `!include` statements are ignored: I set the definition of UefiDriverEntryPoint in MdePkg/MdeLibs.dsc.inc to a clearly invalid path, copied the include statement to the end of all LibraryClasses just in case it might be overridden by UefiPayloadPkg.dsc to the correct value, and the build proceeded as normal.

If nobody else has any comments, then I think this is the last thing to address.

SergiiDmytruk commented:
Yes, I can see that too.  This didn't happen before and everything works properly in a fork on top of `edk2-stable202502`, so something must have changed since that tag.  I'll try to bisect commits to find the cause of the regression (doesn't look like an intended change to me).

benjamindoron commented on UefiPayloadPkg/BlSupportDxe/BlSupportDxe.h:
```diff
@@ -16,12 +16,16 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/DxeServicesTableLib.h>
 #include <Library/DebugLib.h>
 #include <Library/BaseMemoryLib.h>
+#include <Library/BlParseLib.h>
```
No longer needed. Which isn't the end of the world, but https://github.com/tianocore/edk2/pull/11294 will be in conflict with each other anyways

SergiiDmytruk replied:
Removed: https://github.com/tianocore/edk2/compare/d814b557bf198d5d85dbfcfae53c8d451b0b3918..8c20b925ba18c22e663dff13f4d283fb18c51aa4#diff-bb81dd4d85f014adffbae8c7cceb45c3103ff78ad7a22e3d6bab89744fa9a89f

gdong1 reviewed (approved):
This change looks good to me in general. BTW, BlParseLib is used only for legacy UEFI payload (using FV/FD image format). Now universal payload ( ELF or FIT image format) is preferred.

benjamindoron commented:
> > Yes, I can see that too. This didn't happen before and everything works properly in a fork on top of `edk2-stable202502`, so something must have changed since that tag. I'll try to bisect commits to find the cause of the regression (doesn't look like an intended change to me).
> 
> Figured it out and made a fix in [d16b065](https://github.com/tianocore/edk2/commit/d16b0658bc2d87c6a66e7945c4ad4021818d6cc9) which contains much more details. The fix can be moved to a separate PR, although don't know if anything outside of this PR triggers the bug.

Ah, neat! That looks better, thanks. I think I can add a push label now, and if the BaseTools maintainers want changes, they can request them, which will hold the patch. But we should get this in otherwise.

> > This change looks good to me in general. BTW, BlParseLib is used only for legacy UEFI payload (using FV/FD image format). Now universal payload ( ELF or FIT image format) is preferred.
> 
> As far as I understand Universal Payload just won't work with coreboot at the moment when it comes to unmodified upstreams. This repository needs to either merge #6382 and coreboot should get [its part](https://review.coreboot.org/c/coreboot/+/85643) or https://github.com/StarLabsLtd/edk2/tree/universalpayload/UefiPayloadPkg/ShimLayer needs to be added here (coreboot already has the needed changes). So this is something for the future.

I believe SBL might be using UPL already, but yes, coreboot isn't quite ready to switch. I have another branch for coreboot testing (https://github.com/benjamindoron/edk2/tree/upl/coreboot_node_sizes), since we use 32-bit `size` values, and the current code hard-codes for 64-bit. Doing that dynamically is a to-do of mine sometime. From a feature perspective, coreboot isn't ready to switch either: SBL was ready, since they use HOBs and have a `/upl-custom/hoblistptr` property to pass them across, but coreboot's UPL FDT isn't hooked into the entire codebase, and generic drivers (like SMMSTORE, CFR, payload MM) don't produce it, only COREBOOT tables. I believe that while upstream has deprecated legacy UefiPayload, it still has users and maintainers, and so long as a legacy change doesn't break UPL, we'll be continuing to maintain both for a little while longer

gdong1 commented:
> I believe SBL might be using UPL already, but yes, coreboot isn't quite ready to switch. I have another branch for coreboot testing (https://github.com/benjamindoron/edk2/tree/upl/coreboot_node_sizes), since we use 32-bit size values, and the current code hard-codes for 64-bit. Doing that dynamically is a to-do of mine sometime. From a feature perspective, coreboot isn't ready to switch either: SBL was ready, since they use HOBs and have a /upl-custom/hoblistptr property to pass them across, but coreboot's UPL FDT isn't hooked into the entire codebase, and generic drivers (like SMMSTORE, CFR, payload MM) don't produce it, only COREBOOT tables. I believe that while upstream has deprecated legacy UefiPayload, it still has users and maintainers, and so long as a legacy change doesn't break UPL, we'll be continuing to maintain both for a little while longer

You are correct—SBL has been using UPL (ELF image format) for several years and also supports FDT-based UPL. SBL can pass bootloader information using either HOBs or FDT. I understand that coreboot’s UPL support is still a work in progress. From the SBL perspective, it’s fine to update the legacy implementation as long as there are no regression issues for UPL or the legacy UEFI payload.
